### ISSUE TITLE
get_latest_training_loss returns 0

### ISSUE THREAD

**Issue by ginward** [RaRe-Technologies/gensim#3228]
#### Problem description

It seems that the `get_latest_training_loss` function in `fasttext` returns only 0: 

```
class callback(CallbackAny2Vec):
    '''Callback to print loss after each epoch.'''

    def __init__(self):
        self.epoch = 0

    def on_epoch_end(self, model):
        loss = model.get_latest_training_loss()
        print('Loss after epoch {}: {}'.format(self.epoch, loss))
        self.epoch += 1

from pprint import pprint as print
from gensim.models.fasttext import FastText
from gensim.test.utils import datapath

# Set file names for train and test data
corpus_file = datapath('lee_background.cor')

model = FastText(vector_size=100, callbacks=[callback()])

# build the vocabulary
model.build_vocab(corpus_file=corpus_file)

# train the model
model.train(
    corpus_file=corpus_file, epochs=model.epochs,
    total_examples=model.corpus_count, total_words=model.corpus_total_words,
    callbacks=model.callbacks, compute_loss=True,
)

print(model)
```

```
'Loss after epoch 0: 0.0'
'Loss after epoch 1: 0.0'
'Loss after epoch 2: 0.0'
'Loss after epoch 3: 0.0'
'Loss after epoch 4: 0.0'
```

#### Versions

Please provide the output of:

```python
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import struct; print("Bits", 8 * struct.calcsize("P"))
import numpy; print("NumPy", numpy.__version__)
import scipy; print("SciPy", scipy.__version__)
import gensim; print("gensim", gensim.__version__)
from gensim.models import word2vec;print("FAST_VERSION", word2vec.FAST_VERSION)
```

**Comment by ginward** [comment]
This is related to #2658, which probably should not be closed.
It seems that currently fasttext would not return the correct loss using `get_latest_training_loss`.

**Comment by gojomo** [comment]
#2658 is closed as a duplicate, because #2617 is a more comprehensive discussion of what broken (or simply never implemented) in the *2Vec models.

The docs are wrong to imply there's any loss-tallying in `FastText` - it's never been implemented. That could be corrected right away, by overriding the superclass method with another that documents/warns that there's no loss-tracking yet for the `FastText` model. Actually adding loss-trackiing to `FastText` (and `Doc2Vec`) will require a bit more design & work, as hinted in #2617 (& some of the other issues it references).

**Comment by ginward** [comment]
> #2658 is closed as a duplicate, because #2617 is a more comprehensive discussion of what broken (or simply never implemented) in the *2Vec models.
> 
> The docs are wrong to imply there's any loss-tallying in `FastText` - it's never been implemented. That could be corrected right away, by overriding the superclass method with another that documents/warns that there's no loss-tracking yet for the `FastText` model. Actually adding loss-trackiing to `FastText` (and `Doc2Vec`) will require a bit more design & work, as hinted in #2617 (& some of the other issues it references).

I see. But if loss training has never been implemented, how do we know if the training needs to be early stopped, or if the training needs more epochs?

**Comment by gojomo** [comment]
> I see. But if loss training has never been implemented, how do we know if the training needs to be early stopped, or if the training needs more epochs?

You'd have to use other heuristics. AFAIK, neither the original Google `word2vec.c` (on which Gensim's original implementation of the word2vec algorithm was closely based) nor the Facebook `fasttext` tool even offer early-stopping as an option: you pick you `epochs` & live with it, and if you later suspect it was too little or too much, you try another value in a separate run. 

They do each, however, show a running loss that a user can watch for hints.

It's definitely a desirable feature to have - hence the many requests, & partial/buggy implementation inside Gensim's `Word2Vec`, & the open #2617 expressing a goal of fixing/completing the work! It's just not been done, or urgently-required by someone who was sufficiently skilled & motivated to contribute/fund the necessary work, *yet*.

(Note, though, it's also somewhat prone to misinterpretation, with some people thinking it's an accurate measure of model quality for *other* purposes, and that, of a set of candidate models, the one with the lowest loss will work best for outside purposes. That's not inherently the case, as it's just a report on the model's *internal* training goal, which is at best if all sorts of othr things are done right an approximation of the *external* purposes for which people want word-vectors. For example, a massively-'overfit' model can have an arbitrarily low training loss, while being entirely useless for other tasks.)

**Comment by cpuodzius** [comment]
After reading out some replies here and on stackoverflow, I'm aware that loss-tallying is yet to be implemented.
However, the running loss after each epoch - differently than said [here](https://github.com/RaRe-Technologies/gensim/issues/3228#issuecomment-917119337)- is also always zero to me.

I'm running `gemsim==4.0.1` and my example code is:

```
from gensim.models.callbacks import CallbackAny2Vec

class LossLogger(CallbackAny2Vec):
    '''Callback to print loss after each epoch.'''

    def __init__(self):
        self.epoch = 0

    def on_epoch_end(self, model):
        loss = model.get_latest_training_loss()
        print('Loss after epoch {}: {}'.format(self.epoch, loss))
        self.epoch += 1

callbacks = [LossLogger()]
        
doc2vec_model = Doc2Vec(
            documents,
            vector_size=128,
            window=0,
            min_count=5,
            dm=0,
            sample=0.0001,
            workers=4,
            epochs=10,
            alpha=0.025,
            seed=42,
            compute_loss = True,
            callbacks = callbacks
        )
```

```Loss after epoch 0: 0.0
Loss after epoch 1: 0.0
Loss after epoch 2: 0.0
Loss after epoch 3: 0.0
Loss after epoch 4: 0.0
Loss after epoch 5: 0.0
Loss after epoch 6: 0.0
Loss after epoch 7: 0.0
Loss after epoch 8: 0.0
Loss after epoch 9: 0.0
```

Why the `model.get_latest_training_loss()` return always 0 even if it was initialized with `True` ?

**Comment by gojomo** [comment]
Gensim *2Vec model oss-tallying is...
...in `Word2Vec`, buggy (but somewhat usable w/ workarounds).
...in `Doc2Vec`, never yet implemented, hence always 0.
...in `FastText`, never yet implemented, hence always 0.

**Comment by mpenkov** [comment]
Shouldn't we raise NotImplementedError instead of returning zero? It'd be less surprising for the user.

**Comment by gojomo** [comment]
> Shouldn't we raise NotImplementedError instead of returning zero? It'd be less surprising for the user.

That'd be better than the current mysteriously-incomplete behavior! But such hard failures should start as soon as the user takes any step guaranteed to disappoint - such as initializing a model that can't track loss with `compute_loss=True`.  And of course the fasttext doc-comments also shouldn't be describing `compute_loss` and `get_latest_training_lost()` as if they were functional while they're not.
